src/domains/queries.ml

```diff
@@ -37,7 +37,32 @@ end
 module LS = SetDomain.ToppedSet (Lval.CilLval) (struct let topname = "All" end)
 module TS = SetDomain.ToppedSet (CilType.Typ) (struct let topname = "All" end)
 module ES = SetDomain.Reverse (SetDomain.ToppedSet (CilType.Exp) (struct let topname = "All" end))
-module LiftedExp = Lattice.Flat(CilType.Exp)(struct let top_name = "Top" let bot_name = "Unreachable" end)
+module LiftedExp =
+struct
+  include Lattice.Flat(CilType.Exp)(struct let top_name = "Top" let bot_name = "Unreachable" end)
+  let to_invariant = function
+    | `Lifted e -> Some e
+    | `Bot | `Top -> None
+  let of_invariant = function
+    | Some e -> `Lifted e
+    | None -> `Top (* TODO: ??? *)
+
+  let join x y =
+    match (x,y) with
+    | (`Top, _) -> `Top
+    | (_, `Top) -> `Top
+    | (`Bot, x) -> x
+    | (x, `Bot) -> x
```

Comment:
delegate all cases except the `Lifted` `Lifted` one to `join`?

Reply:
When prototyping the Apron domain lifter, which compares with box to output only non-box invariants, I realized that I'll have to rework this here, because of the disjunction handling w.r.t. paths.
When one path doesn't provide an invariant (`Invariant.none`), it essentially means `true`, but another path does provide an invariant, then `Invariant.(||)` will currently only leave the second path's expression, which doesn't hold in the first path, so is unsound.

I'm going to merge the lifted expression lattice here into `Invariant` itself, such that the top and bottom invariants are properly distinguished there. Because as-is, `Invariant.none` corresponds to both a tautology and contradiction, which is a contradiction in itself! Essentially, `Invariant` should then be a fully symbolic lattice, but doesn't implement `leq` since that requires entailment checking. But `leq` isn't needed on these invariants anyway.